File: backend/python/app/middlewares/session_middleware.py
from flask import request, g, Flask
import time
from sqlalchemy.orm.exc import NoResultFound
from app.utils.response_error import ForbiddenError
from app.models import Session


def session_middleware(app: Flask):
    @app.before_request
    def handle_session():
        try:
            session_id = request.headers.get("x-session-id")
            if not session_id:
                return

            session = Session.query.filter_by(id=session_id).first()

            if session and session.user:
                if session.expires_at < int(time.time()):
                    raise ForbiddenError("Session expired")

                g.user = {
                    "id": session.user.id,
                    "name": session.user.name,
                    "email": session.user.email,
                    "phone_number": session.user.phone_number,
                    "bio": session.user.bio,
                }
        except NoResultFound:
            pass
        except Exception as e:
            raise e

File: backend/nodejs-typescript-graphql/src/middlewares/session.middleware.ts
import { NextFunction, Request, Response } from "express";
import z from "zod";
import Session from "@/models/session";
import moment from "@/libs/moment";
import { ResponseError } from "@/utils/response-error.util";

export default async function sessionMiddleware(
  req: Request,
  _: Response,
  next: NextFunction,
) {
  try {
    const sessionId = z
      .string()
      .uuid()
      .safeParse(req.headers["x-session-id"]).data;

    if (!sessionId) {
      next();
      return;
    }

    const session = await Session.query()
      .withGraphFetched("user")
      .where("id", sessionId)
      .first();

    if (session && session.user) {
      if (session.expires_at < moment().unix()) {
        throw new ResponseError(403, "Session expired");
      }

      req.user = {
        id: session.user.id,
        name: session.user.name,
        email: session.user.email,
        phoneNumber: session.user.phone_number,
        bio: session.user.bio,
      };
    }
  } catch (error) {
    next(error);
  }

  next();
}
